internal/pkg/service/buffer/stats/manager.go

```diff
@@ -0,0 +1,104 @@
+package stats
+
+import (
+	"context"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/key"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/store/model"
+)
+
+type Manager struct {
+	ch       chan update
+	syncFn   syncFn
+	clock    clock.Clock
+	ticker   *clock.Ticker
+	perSlice map[key.SliceStatsKey]*stats
+}
+
+type syncFn func(context.Context, []model.SliceStats)
+
+type update struct {
+	key            key.SliceStatsKey
+	size           uint64
+	lastReceivedAt time.Time
+}
+
+type stats struct {
+	count          uint64
+	size           uint64
+	lastReceivedAt time.Time
+	changed        bool
+}
+
+func New(store *store.Store) Manager {
+	clock := clock.New()
+	return Manager{
+		// channel needs to be large enough to not block under average load
+		ch:       make(chan update, 2048),
+		syncFn:   syncToStore(store),
+		clock:    clock,
+		ticker:   clock.Ticker(time.Second),
+		perSlice: make(map[key.SliceStatsKey]*stats),
+	}
+}
+
+func (m *Manager) Notify(key key.SliceStatsKey, size uint64) {
+	m.ch <- update{
+		key:            key,
+		size:           size,
+		lastReceivedAt: m.clock.Now(),
+	}
+}
+
+func (m *Manager) Watch(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				m.ticker.Stop()
+				return
+			case update := <-m.ch:
+				m.handleUpdate(update)
+			case <-m.ticker.C:
+				m.handleSync(ctx)
+			}
+		}
+	}()
+}
+
+func (m *Manager) handleUpdate(update update) {
+	// Init stats
+	if _, exists := m.perSlice[update.key]; !exists {
+		m.perSlice[update.key] = &stats{}
+	}
+
+	// Update stats
+	stats := m.perSlice[update.key]
+	stats.count += 1
+	stats.size += update.size
+	if update.lastReceivedAt.After(stats.lastReceivedAt) {
+		stats.lastReceivedAt = update.lastReceivedAt
+	}
+	stats.changed = true
```

Comment:
S tym `channel` je to pekne :+1:

Reply:
Obavam sa, ale ze to bude v runtime padat, lebo naraz do:
- `m.perSlice`
- `stats.count/size/lastReceivedAt`

Zapisujes aj citas. Channel ti elegantne vyriesil paralelny zapis.
No aj tak budes potrebovat `RWMutex` (mozno to vysledne bude teda bez channel jednoduchsie, uvidis).

`Since Golang 1.6, map cannot be read when it's being written`
https://stackoverflow.com/a/38140573

Je tam ^^^ aj priklad testu.

Reply:
Mohol by si prosim upravit test aby fungoval paralelne.
Paralelne chodia `Notify` a raz za cas sa robi `Sync`.
`go test -race`, ktory pouzivame, by mal odhalit tento problem, ked sa test upravi.

Reply:
Ak by si nevedel ako na to, tak napis.

Reply:
Nechápu. Do `perSlice` sahá jen ta jedna goroutina, která se spouští ve `Watch`, takže tam žádný data race není.

Ten test už paralelní je, volá se tam `Watch`, který teda běží paralelně s calls do `Notify`. Každý call do mocked `clock.Add` má v sobě `time.Sleep`, takže je tam yield, při kterém běží ten sync, takže ta `Watch` goroutine má prostor na to, aby provedla sync.

Můžeš víc popsat to, co myslíš tím "upravit test, aby fungoval paralelně"?

Reply:
Ok, popisem to lepsie, daj mi chvilku.

Reply:
Pozrel som sa na to poriadne.

Nedoslo mi, ze ty to mas v jednej `for` loop, teda sa NEmoze stat, ze sa naraz do map zapisuje aj cita.
Ja som o tom  rozmyslal ako o 2 paralelnych operaciach, sorry.
https://github.com/keboola/keboola-as-code/blob/138fdd647c92afa8ee6a7917242b6b38efd33f00/internal/pkg/service/buffer/stats/manager.go#L59-L69

V tom pripade to ma ale ten problem, ze to obsluhuje jedna goroutine.

Problem je to myslim hlavne tu, kedze je to blokujuca operacia do etcd:
https://github.com/keboola/keboola-as-code/blob/138fdd647c92afa8ee6a7917242b6b38efd33f00/internal/pkg/service/buffer/stats/manager.go#L97

Ale to vies jednoducho obalit do goroutiny.

Idem spravit cele review.